pyquil/quilbase.py

```diff
@@ -1379,33 +1429,62 @@ def get_qubits(self, indices: bool = True) -> Set[QubitDesignator]:
         return _get_frame_qubits(self.frame, indices)
 
 
-class DelayFrames(AbstractInstruction):
-    def __init__(self, frames: List[Frame], duration: float):
-        # all frames should be on the same qubits
-        if len(frames) == 0:
-            raise ValueError("DELAY expected nonempty list of frames.")
-        if len(set(tuple(f.qubits) for f in frames)) != 1:
-            raise ValueError("DELAY with explicit frames requires all frames are on the same qubits.")
-
-        self.frames = frames
-        self.duration = duration
+class Delay(quil_rs.Delay, AbstractInstruction):
+    def __new__(cls, frames: List[Frame], qubits: List[Union[int, Qubit, FormalArgument]], duration: float) -> "Delay":
+        frame_names = [frame.name for frame in frames]
+        rs_qubits = _convert_to_rs_qubits(Delay._join_frame_qubits(frames, qubits))
+        expression = quil_rs_expr.Expression.from_number(complex(duration))
+        return super().__new__(cls, expression, frame_names, rs_qubits)
 
     def out(self) -> str:
-        qubits = self.frames[0].qubits
-        ret = "DELAY " + _format_qubits_str(qubits)
-        for f in self.frames:
-            ret += f' "{f.name}"'
-        ret += f" {self.duration}"
-        return ret
+        return str(self)
 
+    @staticmethod
+    def _join_frame_qubits(
+        frames: List[Frame], qubits: List[Union[int, Qubit, FormalArgument]]
+    ) -> List[Union[int, Qubit, FormalArgument]]:
+        merged_qubits = set(qubits)
+        for frame in frames:
+            merged_qubits.update(frame.qubits)  # type: ignore
+        return list(qubits)
```

Comment:
In `quil-rs`, `Delay` is defined as a `duration`, a vector of frame _name_ `String`s, and a list of `Qubit`s. This is different from `DelayFrames` in pyQuil, which is defined by `List[Frame]`, where each `Frame` can have a list of `Qubit`s associated with it. There is no way to store this association on a `quil-rs` `Delay`. This means we have to do some merging here, but it also means we lose data and won't have exactly the same functionality as before. 

(1) When getting a `Delay`s frames, we can't know which qubits were associated with which, so we can only return them a list of `Frames` with only their names.

(2) When setting a `Delay`s frames, we can't remove the qubits associated with the previously defined frames, because we don't know what they were

(3) When setting a `Delay`s qubits, we will wipe out any qubits that were associated with a `Frame`.

What should we do to reconcile? Deprecate and document, or update quil-rs? I believe `quil-rs` more faithfully follows [the spec](https://quil-lang.github.io/#12-7Timing-and-Synchronization) here, in that the DELAY instruction is defined as:

`DELAY <Qubit>+ <FrameIdentifier>* <Expression>`

Reply:
The qubits in each of the `DelayFrames`'s `Frame`s are the same qubits in the `Delay`s list of qubits.

> There is no way to store this association on a quil-rs Delay.

The association is implied: to get all  `Frame`s from a (quil-rs) `Delay`, you would do something like
```
for frame_name in frame.names:
    build a pyQuil Frame with name=frame_name and qubits=frame.qubits
```

Reply:
> (2) When setting a Delays frames, we can't remove the qubits associated with the previously defined frames, because we don't know what they were

If you're adding a frame, then add the associated qubits to the `Delay`s qubits (treated as a set, no duplicates).

If you're removing a frame, just remove the frame name, since all other frames still require (by definition) the same set of qubits.

Reply:
> (3) When setting a Delays qubits, we will wipe out any qubits that were associated with a Frame.

Yeah, this one is a little iffy. I suppose it could be the case where you add a qubit that doesn't have one of the associated frames. I don't think there's any way for quil-rs to know about that association, so either:
* disallow adding qubits
* allow it, but document that it is unsafe/unchecked